### PR TITLE
Use proper codec in client

### PR DESCRIPTION
**Pull request by wojtek-t**

### CODE REVIEW

File: pkg/client/restclient/request.go [hunk 1]
```diff
@@ -517,6 +517,20 @@ func (r *Request) Timeout(d time.Duration) *Request {
 	return r
 }
 
+func (r *Request) encoder() (runtime.Encoder, bool) {
+	negotiated := r.content.NegotiatedSerializer
```
Comment (smarterclayton) [inline pkg/client/restclient/request.go]: I think we need to be doing this during client initialization.  If the client fails to start, I'd rather error there than here.  Do we really expect to be changing codecs mid client?

File: pkg/client/restclient/request.go [hunk 2]
```diff
@@ -548,6 +549,24 @@ func (r *Request) decoder(contentType string) (runtime.Decoder, bool) {
 	return negotiated.DecoderToVersion(serializer, internalGV), true
 }
 
+func (r *Request) streamingDecoder(contentType string, reader io.ReadCloser) (streaming.Decoder, runtime.Decoder, bool) {
```
Comment (smarterclayton) [inline pkg/client/restclient/request.go]: Although this is better than doing it up front (since most clients won't watch) I do think we should memoize this some how.  We need to get a codec if one is not set, but we don't really support dynamically changing content-type afterwards.  I guess we have a few cases where we might want to speculatively use one type, and then fall back to another, but it's rare.

File: pkg/client/restclient/request.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/client/restclient/request.go]: @smarterclayton - I'm still trying to have this working and I still didn't think about performance of that solution. I will get back to it, once I have tests passing.

File: pkg/watch/versioned/decoder.go [hunk 3]
```diff
@@ -14,53 +14,59 @@ See the License for the specific language governing permissions and
...
 	"encoding/json"
 	"fmt"
 	"io"
 
 	"k8s.io/kubernetes/pkg/runtime"
+//	"k8s.io/kubernetes/pkg/runtime/serializer/streaming"
 	"k8s.io/kubernetes/pkg/watch"
 )
 
 // Decoder implements the watch.Decoder interface for io.ReadClosers that
```
Comment (smarterclayton) [inline pkg/watch/versioned/decoder.go]: I might suggest making this not depend on versioned - instead, you can depend on the internal watch type and do a decode into (the local streaming serializer should handle converting versioned.Event to internal.Event)

**Comment by smarterclayton** [comment]
I've validated at least that we get correct data from the server, back to the client, and that the core framing and decoding code works correctly (https://github.com/kubernetes/kubernetes/pull/24932/commits/cc4d2620e29a7bde3e321c484cb4d996af5abff0) in a test case.  Which probably leaves a couple of things - the embedded serializer is wrong (not JSON?) or some aspect of the client configuration is not actually returning the correct serializer.

File: pkg/watch/versioned/decoder.go [hunk 4]
```diff
@@ -14,56 +14,54 @@ See the License for the specific language governing permissions and
...
 
-// Decode blocks until it can return the next object in the writer. Returns an error
-// if the writer is closed or an object can't be decoded.
+// Decode blocks until it can return the next object in the reader. Returns an error
+// if the reader is closed or an object can't be decoded.
 func (d *Decoder) Decode() (watch.EventType, runtime.Object, error) {
-	var got WatchEvent
-	if err := d.decoder.Decode(&got); err != nil {
+	var got Event
+	if _, _, err := d.decoder.Decode(nil, &got); err != nil {
```
Comment (smarterclayton) [inline pkg/watch/versioned/decoder.go]: use `runtime.DecodeInto(d.decoder, &got)` to check that the object you're decoding into can actually be decoded into.

**Comment by smarterclayton** [comment]
Verified that the decoder implementation is correct for JSON when instantiated directly.

**Comment by wojtek-t** [comment]
Hmm - I will try to debug more. Thanks for debugging @smarterclayton !

File: pkg/watch/versioned/decoder.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/watch/versioned/decoder.go]: We can't use DecodeInfo, because "decoder" is streaming.Decoder (which has different interface than Decoder).

Comment (wojtek-t) [inline pkg/watch/versioned/decoder.go]: However, I'm passing the "into" (got) as an argument and it is supposed to check if type is ok.

**Comment by wojtek-t** [comment]
>
I've validated at least that we get correct data from the server, back to the client, and that the core framing and decoding code works correctly (cc4d262) in a test case. Which probably leaves a couple of things - the embedded serializer is wrong (not JSON?) or some aspect of the client configuration is not actually returning the correct serializer.

I added some more logs to validate failure, and it's definitely not embedded serializer, because we are not even reaching it - I'm seeing a number of such logs:

    E0429 07:45:27.039123       7 decoder.go:51] streaming decoder failed

That said this is either framer or streaming serializer blame for it (either wrong or misconfigured).

File: pkg/watch/versioned/decoder.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/watch/versioned/decoder.go]: You need to check and use the returned object - the contract of decode is
that the returned object might be different from the into object.

Comment (wojtek-t) [inline pkg/watch/versioned/decoder.go]: done

**Comment by wojtek-t** [comment]
Since tests seems to be passing now, I'm reassinging it to you @smarterclayton for review. We should fix the "buffer growing" issue first before merging it, but this is mostly ready for review.

**Comment by smarterclayton** [comment]
Reproduced with test https://github.com/kubernetes/kubernetes/pull/24932/commits/e43e55e2d21cdd41eaf85e4a641804357b1fa303

Going to fix now.

**Comment by smarterclayton** [comment]
JSONFramedReader was reading too much data (past length of buffer), so the base64 encoding of a very long secret was screwed up, which resulted in the resulting value not being a valid base64 object.  Fixing

**Comment by smarterclayton** [comment]
Fix is in https://github.com/kubernetes/kubernetes/pull/24932/commits/b3fca0ad06d28bb4728ca95ca8f90207ab3c1c62 with a test that reproduces if you want to cherrypick it here and we can cross review.

**Comment by wojtek-t** [comment]
@smarterclayton - I've added some comments on that commit.
However, since I will be OOO on Mon and Tue I don't want to block your fix on my PR. So I think it makes sense to have it in a separate PR. WDYT?

File: pkg/client/restclient/config.go [hunk 5]
```diff
@@ -130,9 +130,15 @@ type ContentConfig struct {
...
 	GroupVersion *unversioned.GroupVersion
+	// NegotiatedSerializer is used for obtaining encoders and decoders for multiple
+	// supported media types.
+	NegotiatedSerializer runtime.NegotiatedSerializer
+
 	// Codec specifies the encoding and decoding behavior for runtime.Objects passed
 	// to a RESTClient or Client. Required when initializing a RESTClient, optional
 	// when initializing a Client.
+	//
+	// DEPRECATED: Please use NegotiatedSerializer instead.
```
Comment (smarterclayton) [inline pkg/client/restclient/config.go]: If this takes precedence (or does not) the doc should specify its relationship to NegotiatedSerializer.  I suspect there are still cases where you want to pass a codec and not do negotiation, so I don't know that this is deprecated.

File: pkg/client/restclient/client.go [hunk 6]
```diff
@@ -119,6 +135,33 @@ func readExpBackoffConfig() BackoffManager {
...
+func createSerializers(config ContentConfig) (*Serializers, error) {
+	negotiated := config.NegotiatedSerializer
+	contentType := config.ContentType
+	serializer, ok := negotiated.SerializerForMediaType(contentType, nil)
+	if !ok {
+		return nil, fmt.Errorf("Serializer for %s not registered", contentType)
+	}
+	streamingSerializer, framer, _, ok := negotiated.StreamingSerializerForMediaType(contentType, nil)
+	if !ok {
+		return nil, fmt.Errorf("Streaming serializer for %s not registered", contentType)
```
Comment (smarterclayton) [inline pkg/client/restclient/client.go]: errors should be lowercased

File: pkg/client/restclient/negotiated_codec.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/runtime"
+)
+
+// FIXME: This required some work.
+// FIXME: This should use versioning.NewCodecForScheme to get encoder and decoder.
+// TODO: This should probably be renamed.
+type negotiatedSerializerWrapper struct {
```
Comment (smarterclayton) [inline pkg/client/restclient/negotiated_codec.go]: This probably belongs in the runtime/serializer package as a component, and we should figure out what happens when someone asks encoder for version and it conflicts with the raw serializer.  That's possibly another reason to continue accepting codec for now.

File: pkg/client/restclient/request.go [hunk 8]
```diff
@@ -670,7 +673,9 @@ func (r *Request) Watch() (watch.Interface, error) {
 		}
 		return nil, fmt.Errorf("for request '%+v', got status: %v", url, resp.StatusCode)
 	}
-	return watch.NewStreamWatcher(watchjson.NewDecoder(resp.Body, r.content.Codec)), nil
+	framer := r.serializers.Framer.NewFrameReader(resp.Body)
+	decoder := streaming.NewDecoder(framer, r.serializers.StreamingSerializer)
```
Comment (smarterclayton) [inline pkg/client/restclient/request.go]: Do you need to defer a Close() on the decoder?

**Comment by smarterclayton** [comment]
A few comments, everything else looks ok to me.

File: pkg/client/restclient/config.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/client/restclient/config.go]: Codec is currently used only by some tests that I didn't yet migrated to use NegotiatedSerializer. I'm going to do it in a separate PR soon.
Update the comment to reflect that.

File: pkg/client/restclient/client.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/client/restclient/client.go]: done

File: pkg/client/restclient/request.go (hunk 8 above)
Comment (wojtek-t) [inline pkg/client/restclient/request.go]: StreamWatcher is doing that.

File: pkg/client/restclient/negotiated_codec.go (hunk 7 above)
Comment (wojtek-t) [inline pkg/client/restclient/negotiated_codec.go]: Moved to pkg/serializer and added a TODO for conflict issue.

File: pkg/client/restclient/config.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/client/restclient/config.go]: Regarding passing codec - we should pass this "negotiatedSerializerWrapper" in those cases, I think.

**Comment by wojtek-t** [comment]
@smarterclayton - comments addressed PTAL

However, it seems that your fix #25014 didn't fully solve the problem. Without the 3rd commit (increasing the initial buffer size to 128kB), the:
"KubeletManagedEtcHosts should test kubelet managed /etc/hosts file"
is regularly failing.

**Comment by smarterclayton** [comment]
LGTM

File: pkg/client/restclient/client.go [hunk 9]
```diff
@@ -119,6 +135,33 @@ func readExpBackoffConfig() BackoffManager {
...
+	streamingSerializer, framer, _, ok := negotiated.StreamingSerializerForMediaType(contentType, nil)
+	if !ok {
+		return nil, fmt.Errorf("streaming serializer for %s not registered", contentType)
+	}
+	if framer == nil {
+		return nil, fmt.Errorf("no framer for %s", contentType)
+	}
+	internalGV := unversioned.GroupVersion{
+		Group:   config.GroupVersion.Group,
+		Version: runtime.APIVersionInternal,
```
Comment (caesarxuchao) [inline pkg/client/restclient/client.go]: The clientset clients should deocode (actually just deserialize) to versioned objects, so for them the version here should be config.GroupVersion.Version. How about adding another field to the client config to specify the decode version? I can send a PR if you think adding an extra config field is acceptable. @wojtek-t @smarterclayton

Comment (smarterclayton) [inline pkg/client/restclient/client.go]: Clients probably wants no versioning at all - so the wrappers below are wrong.  I think that would be a "non-converting" client, which means that this method would be a lot simpler for them.  That might be better as a boolean that avoids the wrappers.

Comment (caesarxuchao) [inline pkg/client/restclient/client.go]: Yes, actually the clientset should use "non-converting" client.

Comment (smarterclayton) [inline pkg/client/restclient/client.go]: Although it *may* be that you need to ensure kind and Version are set
in the external client.  That's a conversion responsibility today, but
I think a non-converting client would be surprised by that.  I'll
think about it over the weekend and offer and suggestions I come up
with.

We don't want to be accidentally converted in the primary resource
path, which is the most important part.

Comment (caesarxuchao) [inline pkg/client/restclient/client.go]: Opened https://github.com/kubernetes/kubernetes/issues/25589

**Comment by wojtek-t** [comment]
Ref #8132
